### ISSUE TITLE
Write simple TURN and STUN client

### ISSUE THREAD

**Issue by Sean-Der** [pions/turn#15]
A simple TURN/STUN client would be great for testing, currently there isn't a nice portable tool to assert that both are working (most people suggest https://webrtc.github.io/samples/src/content/peerconnection/trickle-ice/)

**Comment by iwittkau** [comment]
How would one go about trying to implement a client? Are there any ressources explaing the protocol?

**Comment by Sean-Der** [comment]
I am so excited you are going to contribute! This is what I would do.

create `cmd/client/main.go` this will be a new STUN client, you should be able to just do `go run cmd/client/main.go`

STUN works by just sending STUN packets over UDP, we have code that packs/unpacks these packets.

You should create a new UDP listener, and then send a STUN packet to any STUN server. You should then get a STUN packet back. I would write in one Go routine (and read in the other) this is pseudo code of what it will look like.

We listen for incoming STUN traffic, and then send a STUN binding request. This STUN binding request is so that we can find out what our public IP is. This is how we process STUN packets now on the server https://github.com/pions/turn/blob/master/internal/server/server.go#L71 you will need something like this to process the response that comes from the server (and then print it)

You should be able to create a neat little utility that will print your public IP. This will really come in handy for testing STUN servers (and can be used for other ones as well!) Feel free to ask any questions on Slack. The only other thing that is useful for learning is the RFC, but it is dense https://tools.ietf.org/html/rfc5389

```
func sendStunRequest(conn *ipv4.PacketConn) {
   	stun.BuildAndSend(connection, YOUR_STUN_SERVER, stun.ClassRequest, stun.MethodBinding, stun.GenerateTransactionId())
}

func main() {
        packet := make([]byte, 1500)
	c, err := net.ListenPacket("udp4", "0.0.0.0:0")
	if err != nil {
		return errors.Wrap(err, fmt.Sprintf("Failed to listen on %s", listeningAddress))
	}
	connection := ipv4.NewPacketConn(c)
	if err := connection.SetControlMessage(ipv4.FlagDst, true); err != nil {
		return errors.Wrap(err, "Failed to SetControlMessage ipv4.FlagDst")
	}
        
        go sendStunRequest(connection)
	for {
		size, cm, addr, err := connection.ReadFrom(packet)
		if err != nil {
			return errors.Wrap(err, "failed to read packet from udp socket")
		}
		srcAddr, err := stun.NewTransportAddr(addr)
		if err != nil {
			return errors.Wrap(err, "failed reading udp addr")
		}

	}
}
```

**Comment by iwittkau** [comment]
That looks doable. I am familiar with the general protocol for. I'll get back to you with some progress.

**Comment by Sean-Der** [comment]
Implemented by #22
